Conceptual document for single file publish option (#20365)
dotnet/docs: docs/core/deploying/single-file.md

---
title: Single file application
description: Learn what single file application is and why you should consider using this application deployment model.
author: lakshanf
ms.author: lakshanf
ms.date: 08/28/2020
---
# Single file deployment and executable

Bundling all application dependent files into a single binary provides an application developer with the attractive option to deploy and distribute the application as a single file. This deployment model has been available since .NET Core 3.0 and has been enhanced in .NET 5.0 where the code is directly executed from the single file in the client machine. On .NET Core 3.0 version, all files were first self-extracted into a temporary location before running. On Windows .NET 5.0, single-file option will not bundle native binaries since a good debugging experience requires these not to be included in the single file currently. Linux will have all files bundled in .NET 5.0.

> Single File deployment is available for both [framework-dependent deployment model](index.md#publish-framework-dependent) and [self-contained application](index.md#publish-self-contained) although we anticipate that most developers will use Single file with client-side [self-contained application](index.md#publish-self-contained) managed applications for .NET Core 5.0. The single file deployment option can be combined with [Trim](trim-self-contained.md) and [ReadyToRun](../tools/dotnet-publish.md) publish options.

Reviewer: I would either omit the "although we anticipate" clause or be more specific about the applicable scenario(s).

> Single File deployment is available for both [framework-dependent deployment model](index.md#publish-framework-dependent) and [self-contained application](index.md#publish-self-contained) although we anticipate that most developers will use Single file with client-side [self-contained application](index.md#publish-self-contained) managed applications for .NET Core 5.0. The single file deployment option can be combined with [Trim](trim-self-contained.md) and [ReadyToRun](../tools/dotnet-publish.md) publish options.

Author: Thanks

> There are some caveats that you need to be aware for single-file use, chief of which is the use path information to locate a file relative to the location of your application. Assembly. Location API will return an empty string which is the default behavior for assemblies loaded from memory. The compiler will give a warning for this API during build time to alert the developer to the specific behavior. If the path to the application directory is needed, AppContext.BaseDirectory, API will return the directory where the AppHost (the single-file bundle itself) resides. Managed C++ applications are not best suited for single-file deployment and we encourage application to be rewritten in C# or other managed language to be single-file compatible.

Reviewer: Something got garbled in the first part of this paragraph -- "...which is the use path information..." isn't clear and "Assembly." stands by itself as if it were a sentence.  Should this be "use of path information" and "The Assembly.Location API"? Would be ideal to make Assembly.Location an xref link.

Same comment for "If the path..." sentence -- add "the" before "AppContext.BaseDirectory API (remove comma before API) and make it an xref link.

"not best suited" --> "not well suited"
"we encourage application to be rewritten in C#" --> "we recommend that you write applications in C#"

Author: Thanks, added these

## Exclude files being embedded in Single File

Certain files can be explicitly excluded from being embedded in the single-file by setting following meta-data:

```xml
<ExcludeFromSingleFile>true</ExcludeFromSingleFile>
```

For example, to place some files in the publish directory but not bundle them in the single-file:

```xml
<ItemGroup>

>     <Content Update="*.xml">

Reviewer 2: csproj usually uses two spaces for indentation

Author: Thanks!

      <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
      <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
    </Content>
  </ItemGroup>

```

## Publish a Single File app - CLI

Publish a single file application using the [dotnet publish](../tools/dotnet-publish.md) command. When you publish your app, set the following properties:

- Publish as a self-contained app for a specific runtime: `-r win-x64`
- Enable trimming: `/p:PublishSingleFile=true`

> The following example publishes an app for Windows as self-contained single file application.

Reviewer: ```suggestion
The following example publishes an app for Windows as a self-contained single file application.
```

```xml
<PropertyGroup>
    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
    <PublishSingleFile>true</PublishSingleFile>
    <SelfContained>true</SelfContained>
</PropertyGroup>
```

> The following example publishes an app for Linux as Framework Dependent single file application..

Reviewer: ```suggestion
The following example publishes an app for Linux as a framework dependent single file application..
```

    If you don't already have a publishing profile, follow the instructions to create one and choose the **Folder** target-type.

01. Choose **Edit**.

    :::image type="content" source="media/trim-self-contained/visual-studio-publish-edit-settings.png" alt-text="Visual studio publish profile with edit button.":::

01. In the **Profile settings** dialog, set the following options:

    - Set **Deployment mode** to **Self-contained**.
    - Set **Target runtime** to the platform you want to publish to.
    - Select **Produce single file**.

    Choose **Save** to save the settings and return to the **Publish** dialog.

    :::image type="content" source="media/single-file/visual-studio-publish-single-file-properties.png" alt-text="Profile settings dialog with deployment mode, target runtime, and single file options highlighted.":::

01. Choose **Publish** to publish your app as a single file.

For more information, see [Publish .NET Core apps with Visual Studio](deploy-with-vs.md).

## Publish a Single File app - Visual Studio for Mac

> Visual Studio for Mac doesn't provide options to publish your app as a single file during publish. You'll need to publish manually by following the instructions from the [Publish a Single File app - CLI](#publish-a-single-file-app---CLI) section. For more information, see [Publish .NET Core apps with .NET Core CLI](deploy-with-cli.md).

Reviewer 2: ```suggestion
Visual Studio for Mac doesn't provide options to publish your app as a single file during publish. You'll need to publish manually by following the instructions from the [Publish a Single File app - CLI](#publish-a-single-file-app---cli) section. For more information, see [Publish .NET Core apps with .NET Core CLI](deploy-with-cli.md).
```

---
title: Single file application
description: Learn what single file application is and why you should consider using this application deployment model.
author: lakshanf
ms.author: lakshanf
ms.date: 08/28/2020
---
# Single file deployment and executable

> Bundling all application-dependent files into a single binary provides an application developer with the attractive option to deploy and distribute the application as a single file. This deployment model has been available since .NET Core 3.0 and has been enhanced in .NET 5.0 where the code is directly executed from the single file in the client machine. On .NET Core 3.0 version, all files were first self-extracted into a temporary location before running.

Reviewer: ```suggestion
Bundling all application-dependent files into a single binary provides an application developer with the attractive option to deploy and distribute the application as a single file. This deployment model has been available since .NET Core 3.0 and has been enhanced in .NET 5.0 where the code is directly executed from the single file in the client machine. On the .NET Core 3.0 version, all files were first self-extracted into a temporary location before running.
```

> Single File deployment is available for both [framework-dependent deployment model](index.md#publish-framework-dependent) and [self-contained application](index.md#publish-self-contained). The size of the single file in a self-contained application will be large since it will include the runtime and the framework libraries. The single file deployment option can be combined with [Trim](trim-self-contained.md) and [ReadyToRun](../tools/dotnet-publish.md) publish options.

Reviewer: ```suggestion
Single File deployment is available for both the [framework-dependent deployment model](index.md#publish-framework-dependent) and [self-contained applications](index.md#publish-self-contained). The size of the single file in a self-contained application will be large since it will include the runtime and the framework libraries. The single file deployment option can be combined with [Trim](trim-self-contained.md) and [ReadyToRun](../tools/dotnet-publish.md) publish options.
```

> There are some caveats that you need to be aware for single-file use, chief of which is the use path information to locate a file relative to the location of your application. [Assembly.Location](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.location?view=netcore-3.1) API will return an empty string, which is the default behavior for assemblies loaded from memory. The compiler will give a warning for this API during build time to alert the developer to the specific behavior. If the path to the application directory is needed, [AppContext.BaseDirectory](https://docs.microsoft.com/en-us/dotnet/api/system.appcontext.basedirectory?view=netcore-3.1) API will return the directory where the AppHost (the single-file bundle itself) resides. Managed C++ applications are not well suited for single-file deployment and we recommend that you write applications in C# to be single-file compatible.

Reviewer: ```suggestion
There are some caveats that you need to be aware for single-file use, chief of which is the use of path information to locate a file relative to the location of your application. The <xref:System.Reflection.Assembly.Location?displayProperty=nameWithType> API will return an empty string, which is the default behavior for assemblies loaded from memory. The compiler will give a warning for this API during build time to alert the developer to the specific behavior. If the path to the application directory is needed, the <xref:System.AppContext.BaseDirectory?displayProperty=nameWithType> API will return the directory where the AppHost (the single-file bundle itself) resides. Managed C++ applications are not well suited for single-file deployment and we recommend that you write applications in C# to be single-file compatible.
```

> The single-file option for [self-contained applications](index.md#publish-self-contained) will not bundle native libraries on macOS and Windows machines for .NET 5.0. This is to ensure a good debugging experience, which requires native files to be excluded from the single file. Linux will have all files bundled in .NET 5.0 and will execute the application from the single file. Windows and macOS machines will have an option to set a flag, `IncludeNativeLibrariesForSelfExtract`, to include native libraries in the single file bundle, but these files will be extracted to a temporary directory in the client machine when the single file application is run.

Reviewer 3: I think this is misleading. Single-file will not bundle native libraries by default anywhere. On Linux we prelink the runtime into the bundle, so only application native libraries are left next to the bundle, on Windows we prelink only the hosting code, so both the runtime and application native libraries are left next to the bundle. The way the paragraph is written seems to suggest that on Linux we will bundle all native libraries - which is not the case.

Reviewer 3: Also the `IncludeNativeLibrariesForSelfExtract` can be used on all platforms...

> [Self-contained application](index.md#publish-self-contained) single-file option will not bundle native libraries on MacOS and Windows machines for .NET 5.0. This is to ensure a good debugging experience and require native files not to be included in the single file. Linux will have all files bundled in .NET 5.0 and will execute the application from single file. Windows and MacOS machines will have an option to set a flag, `IncludeNativeLibrariesForSelfExtract`, to include native libraries in the single file bundle but these files will be extracted to a temporary directory in the client machine when the single file application is run.

Reviewer: ```suggestion
The single-file option for [self-contained applications](index.md#publish-self-contained) will not bundle native libraries on MacOS and Windows machines for .NET 5.0. This is to ensure a good debugging experience, which requires native files to be excluded from the single file. Linux will have all files bundled in .NET 5.0 and will execute the application from the single file. Windows and MacOS machines will have an option to set a flag, `IncludeNativeLibrariesForSelfExtract`, to include native libraries in the single file bundle, but these files will be extracted to a temporary directory in the client machine when the single file application is run.
```

## Exclude files being embedded in Single File

Certain files can be explicitly excluded from being embedded in the single-file by setting following meta-data:

```xml
<ExcludeFromSingleFile>true</ExcludeFromSingleFile>
```

For example, to place some files in the publish directory but not bundle them in the single-file:

```xml
<ItemGroup>

>     <Content Update="*.xml">

Reviewer 3: This is misleading as well - 5.0 bundle will not bundle anything but assemblies. So `*.xml` will not get bundled by default. I think a better sample here would be to specify an assembly by name - so something like "Plugin.dll" or similar.

      <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
      <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
    </Content>
  </ItemGroup>

```

## Publish a Single File app - CLI

Publish a single file application using the [dotnet publish](../tools/dotnet-publish.md) command. When you publish your app, set the following properties:

- Publish as a self-contained app for a specific runtime: `-r win-x64`

> - Enable trimming: `/p:PublishSingleFile=true`

Reviewer 3: ```suggestion
- Publish as single-file: `/p:PublishSingleFile=true`
```

Author: Good catch, thanks!

> The following example publishes an app for Windows as a self-contained single file application.

Reviewer 3: I find this confusing - the header is "CLI", the above bullet points use command line syntax, but the samples use `csproj`/MSBuild syntax. I know that they do the same thing, but lot of people won't make the connection.

There's also difference in how the SDK behaves in these cases. If the properties are specified in the `csproj` they apply always, so:
* Specifying `RuntimeIdentifier` causes `dotnet build` to output into `/bin/Debug/net5.0/win-x64` (not the default `/bin/Debug/net5.0`.
* Specifying `RuntimeIdentifier` causes `dotnet build` to produce self-contained app (unlike the framework dependent which is the default for normal build)

I personally would not show samples where these settings are in the `csproj`. Because:
* When using CLI - the above weird behaviors will occur
* When using VS the same problems as CLI - and in addition if I use the VS UI to publish it will also create weird interactions - for example the VS UI will not see the values specified in the `.csproj` since it only looks the publish.xml

Author: I wanted to encourage users to specify publish settings via a configuration file but I can see that it can be confusing, So changed as you suggested, thanks!

There are some caveats that you need to be aware for single-file use, first of which is the use of path information to locate a file relative to the location of your application. The <xref:System.Reflection.Assembly.Location?displayProperty=nameWithType> API will return an empty string, which is the default behavior for assemblies loaded from memory. The compiler will give a warning for this API during build time to alert the developer to the specific behavior. If the path to the application directory is needed, the <xref:System.AppContext.BaseDirectory?displayProperty=nameWithType> API will return the directory where the AppHost (the single-file bundle itself) resides. Managed C++ applications aren't well suited for single-file deployment and we recommend that you write applications in C# to be single-file compatible.

> Single-file don't bundle native libraries by default. On Linux, we prelink the runtime into the bundle and only application native libraries are deployed to the same directory as the single-file app. On Windows, we prelink only the hosting code and both the runtime and application native libraries are deployed to the same directory as the single-file app. This is to ensure a good debugging experience, which requires native files to be excluded from the single file. There is an option to set a flag, `IncludeNativeLibrariesForSelfExtract`, to include native libraries in the single file bundle, but these files will be extracted to a temporary directory in the client machine when the single file application is run.

Reviewer 2: nit: Should it be "doesn't" instead of "don't"?

Reviewer 4: hahaha whoops.

```suggestion
Single-file doesn't bundle native libraries by default. On Linux, we prelink the runtime into the bundle and only application native libraries are deployed to the same directory as the single-file app. On Windows, we prelink only the hosting code and both the runtime and application native libraries are deployed to the same directory as the single-file app. This is to ensure a good debugging experience, which requires native files to be excluded from the single file. There is an option to set a flag, `IncludeNativeLibrariesForSelfExtract`, to include native libraries in the single file bundle, but these files will be extracted to a temporary directory in the client machine when the single file application is run.
```

---
title: Single file application
description: Learn what single file application is and why you should consider using this application deployment model.
author: lakshanf
ms.author: lakshanf
ms.date: 08/28/2020
---

> # Single file deployment and executable

Reviewer 2: This heading says "executable" (singular), and the toc says "executables" (plural). Was the intended?

Author: Thanks a lot, Youssef!

> Bundling all application-dependent files into a single binary provides an application developer with the attractive option to deploy and distribute the application as a single file. This deployment model has been available since .NET Core 3.0 and has been enhanced in .NET 5.0 where the code is directly executed from the single file in the client machine. In the .NET Core 3.0 version, all files were first self-extracted into a temporary location before running.

Reviewer 4: - This piece doesn't make sense to me in the given context. I would move it to the end describing it in the context of BEFORE and AFTER.

```suggestion
Bundling all application-dependent files into a single binary provides an application developer with the attractive option to deploy and distribute the application as a single file. This deployment model has been available since .NET Core 3.0 and has been enhanced in .NET 5.0. Previously in .NET Core 3.0, when a user runs your single-file app, .NET Core first extracts all files to a temporary directory before running the application. .NET 5.0 improves this experience by directly running the code without the need to extract the files from the app.
```

Single File deployment is available for both the [framework-dependent deployment model](index.md#publish-framework-dependent) and [self-contained applications](index.md#publish-self-contained). The size of the single file in a self-contained application will be large since it will include the runtime and the framework libraries. The single file deployment option can be combined with [ReadyToRun](../tools/dotnet-publish.md) and [Trim (an experimental feature in .NET 5.0)](trim-self-contained.md) publish options.

> There are some caveats that you need to be aware for single-file use, chief of which is the use of path information to locate a file relative to the location of your application. The <xref:System.Reflection.Assembly.Location?displayProperty=nameWithType> API will return an empty string, which is the default behavior for assemblies loaded from memory. The compiler will give a warning for this API during build time to alert the developer to the specific behavior. If the path to the application directory is needed, the <xref:System.AppContext.BaseDirectory?displayProperty=nameWithType> API will return the directory where the AppHost (the single-file bundle itself) resides. Managed C++ applications are not well suited for single-file deployment and we recommend that you write applications in C# to be single-file compatible.

Reviewer 4: - "Chief of which" isn't a common phrase, I changed this to "first of which"
- Use contractions: I changed `are not` to `aren't`.
- ❓ By stating "we recommend that you write applications in C#" are we saying don't write in F# or VB?

```suggestion
There are some caveats that you need to be aware for single-file use, first of which is the use of path information to locate a file relative to the location of your application. The <xref:System.Reflection.Assembly.Location?displayProperty=nameWithType> API will return an empty string, which is the default behavior for assemblies loaded from memory. The compiler will give a warning for this API during build time to alert the developer to the specific behavior. If the path to the application directory is needed, the <xref:System.AppContext.BaseDirectory?displayProperty=nameWithType> API will return the directory where the AppHost (the single-file bundle itself) resides. Managed C++ applications aren't well suited for single-file deployment and we recommend that you write applications in C# to be single-file compatible.
```

> Single-file will not bundle native libraries by default. On Linux, we prelink the runtime into the bundle and only application native libraries are left next to the bundle. On Windows, we prelink only the hosting code and both the runtime and application native libraries are left next to the bundle. This is to ensure a good debugging experience, which requires native files to be excluded from the single file. There is an option to set a flag, `IncludeNativeLibrariesForSelfExtract`, to include native libraries in the single file bundle, but these files will be extracted to a temporary directory in the client machine when the single file application is run.

Reviewer 4: - Changed `will not bundle` to `don't bundle`.
- ❓ What do you mean by native libraries? Are these native libraries that the app is referencing or that the apphost is referencing?
- ❓ I think that left next to the bundle should be described clearer. "Deployed to the same directory as the single-file app" or something?

```suggestion
Single-file don't bundle native libraries by default. On Linux, we prelink the runtime into the bundle and only application native libraries are left next to the bundle. On Windows, we prelink only the hosting code and both the runtime and application native libraries are left next to the bundle. This is to ensure a good debugging experience, which requires native files to be excluded from the single file. There is an option to set a flag, `IncludeNativeLibrariesForSelfExtract`, to include native libraries in the single file bundle, but these files will be extracted to a temporary directory in the client machine when the single file application is run.
```

> ## Exclude files being embedded in Single File

Reviewer 4: - Considering the subject of the article is single-file deployment, I don't think it needs to be restated here.

```suggestion
## Exclude files from being embedded
```

> Certain files can be explicitly excluded from being embedded in the single-file by setting following meta-data:

Reviewer 4: - Metadata not meta-data
- ❓ I'm not sure this is the correct term. Previously you used the wording "option to set a flag xxxxx" but now you say metadata. I think we usually just say "configure a setting"

```suggestion
Certain files can be explicitly excluded from being embedded in the single-file by setting following metadata:
```

```xml
<ExcludeFromSingleFile>true</ExcludeFromSingleFile>
```

For example, to place some files in the publish directory but not bundle them in the single-file:

```xml
<ItemGroup>
    <Content Update="Plugin.dll">
      <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
      <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
    </Content>
  </ItemGroup>

```

> ## Publish a Single File app - CLI

Reviewer 4: - Heading fix

```suggestion
## Publish a single file app - CLI
```

Publish a single file application using the [dotnet publish](../tools/dotnet-publish.md) command. When you publish your app, set the following properties:

- Publish as a self-contained app for a specific runtime: `-r win-x64`

> - Publish as a single-file: `-p:PublishSingleFile=true`

Reviewer 4: - I would break this up into what settings are required for each deployment mode: framework-dependent and self-contained.
- It's not clear that single-file always requires a RID.

Author: Thanks!

The following example publishes an app for Windows as a self-contained single file application.

```dotnet
dotnet publish -r win-x64 -p:PublishSingleFile=true --self-contained true
```

The following example publishes an app for Linux as a framework dependent single file application..

```dotnet
dotnet publish -r linux-x64 -p:PublishSingleFile=true --self-contained false
```

For more information, see [Publish .NET Core apps with .NET Core CLI](deploy-with-cli.md).

## Publish a Single File app - Visual Studio

Visual Studio creates reusable publishing profiles that control how your application is published.

01. On the **Solution Explorer** pane, right-click on the project you want to publish. Select **Publish...**.

    :::image type="content" source="media/trim-self-contained/visual-studio-solution-explorer.png" alt-text="Solution Explorer with a right-click menu highlighting the Publish option.":::

    If you don't already have a publishing profile, follow the instructions to create one and choose the **Folder** target-type.

01. Choose **Edit**.

>     :::image type="content" source="media/trim-self-contained/visual-studio-publish-edit-settings.png" alt-text="Visual studio publish profile with edit button.":::

Reviewer 4: - ⚠️ Don't use images from other articles, copy the image to the appropriate folder for this article.

Reviewer 2: @adegeo FYI, I believe I recently met some code snippets or images referenced from other articles in the repo.

01. In the **Profile settings** dialog, set the following options:

    - Set **Deployment mode** to **Self-contained**.
    - Set **Target runtime** to the platform you want to publish to.
    - Select **Produce single file**.

    Choose **Save** to save the settings and return to the **Publish** dialog.

    :::image type="content" source="media/single-file/visual-studio-publish-single-file-properties.png" alt-text="Profile settings dialog with deployment mode, target runtime, and single file options highlighted.":::

01. Choose **Publish** to publish your app as a single file.

For more information, see [Publish .NET Core apps with Visual Studio](deploy-with-vs.md).

## Publish a Single File app - Visual Studio for Mac

> Visual Studio for Mac doesn't provide options to publish your app as a single file. You'll need to publish manually by following the instructions from the [Publish a Single File app - CLI](#publish-a-single-file-app---cli) section. For more information, see [Publish .NET Core apps with .NET Core CLI](deploy-with-cli.md).

Reviewer 2: ```suggestion
Visual Studio for Mac doesn't provide options to publish your app as a single file. You'll need to publish manually by following the instructions from the [Publish a single file app - CLI](#publish-a-single-file-app---cli) section. For more information, see [Publish .NET Core apps with .NET Core CLI](deploy-with-cli.md).
```

## See also

- [.NET Core application deployment](index.md).
- [Publish .NET Core apps with .NET Core CLI](deploy-with-cli.md).
- [Publish .NET Core apps with Visual Studio](deploy-with-vs.md).

> - [dotnet publish command](../tools/dotnet-publish.md).

Reviewer 2: To prevent localization.

```suggestion
- [`dotnet publish` command](../tools/dotnet-publish.md).
```
